### PR TITLE
feat: stealth textfield prefill feature from query parameters

### LINKED ISSUE #498: Stealth textfield prefill feature from query parameters
If a query parameter is provided to a form URL in one of the following formats:

Server-side URL
- `https://form.gov.sg/<formId>?<fieldId1>=<value1>&<fieldId2>=<value2>...`

Client-side URL (server can't see anything after hashbang)
- `https://form.gov.sg/#!/<formId>?<fieldId1>=<value1>&<fieldId2>=<value2>...`

and if the `fieldIds` are valid mongoose object IDs and refer to a short text field, then prefill and disable editing the corresponding form field on the frontend.

Path and query parameters can be retrieved with the AngularJS [$location](https://docs.angularjs.org/guide/$location) service. Watch for edge cases depending on whether the [hashbang or query parameters come first](https://github.com/angular/angular.js/issues/6172).

### PR DESCRIPTION
**Pull request by shuli-ogp**
## Problem

Closes #498

## Solution
- Retain query params when redirecting from non-hashbang url
- On public form route, extract query params using $location service on frontend to stealth prefill forms

TODO:

- Test that special characters in query param values work correctly and that they are passed correctly to redirect endpoint

## Screenshots

**AFTER**:
With the query param string: `?<fieldId2>=yes&<fieldId4>=hello`
Screenshot from 2020-10-27 16-29-33

## Tests

- [ ] Create a form with short text field. Open in public view with the query params `<shortTextFieldId>=<value1>`, where `<value1>` comprises alphanumeric/hyphen/underscore characters. Check that the short text field is prefilled with value1
- [ ] Add a duplicate query param `<shortTextFieldId>=<value1a>`. Check that the short text field is not prefilled with any value
- [ ] Add a long text field. Add the query params `<otherFieldId>=<value2>`. Check that the long text field is not prefilled
- [ ] Check that prefilling works even without the hashbang in the url
- [ ] Add radio button to the form. Set form logic to hide the short text field conditional on the radio button. Check that the short text field is prefilled when it is unhidden

### CODE REVIEW

File: src/app/controllers/public-forms.server.controller.js [hunk 1]
```diff
@@ -44,9 +44,19 @@ exports.isFormPublic = function (req, res, next) {
  * @param  {Object} res - Express response object
  */
 exports.redirect = async function (req, res) {
+  const queryPortion = req.url ? req.url.split('?', 2)[1] : undefined
```
Comment (liangyuanruo) [inline src/app/controllers/public-forms.server.controller.js:47]: how about using `req.query` directly?

File: src/app/controllers/public-forms.server.controller.js [hunk 2]
```diff
@@ -44,9 +44,19 @@ exports.isFormPublic = function (req, res, next) {
...
   let redirectPath = req.params.state
     ? req.params.Id + '/' + req.params.state
     : req.params.Id
+  if (
+    queryPortion &&
+    queryPortion.length > 0 &&
+    queryPortion.match(
+      /^(([0-9A-Za-z]+=[0-9A-Za-z]+)(&[0-9A-Za-z]+=[0-9A-Za-z]+)*)?$/,
```
Comment (liangyuanruo) [inline src/app/controllers/public-forms.server.controller.js:55]: since this is a redirect endpoint, perhaps we can simplify the abstraction and forward all query parameters without checks

File: src/app/routes/frontend.server.routes.js [hunk 3]
```diff
@@ -14,7 +14,9 @@ module.exports = function (app) {
     celebrate({
       [Segments.QUERY]: {
         redirectPath: Joi.string()
-          .regex(/^[a-fA-F0-9]{24}(\/(preview|template|use-template))?$/)
+          .regex(
+            /^[a-fA-F0-9]{24}(\/(preview|template|use-template))?(\?([0-9A-Za-z]+=[0-9A-Za-z]+)(&[0-9A-Za-z]+=[0-9A-Za-z]+)*)?$/,
```
Comment (liangyuanruo) [inline src/app/routes/frontend.server.routes.js:18]: ideally, there would be a way to rewrite this so that we don't have to deal with a monster regex

File: src/public/modules/forms/base/directives/field.client.directive.js [hunk 4]
```diff
@@ -1,12 +1,13 @@
 'use strict'
 
 const { get } = require('lodash')
+// const queryString = require('query-string')
```
Comment (liangyuanruo) [inline src/public/modules/forms/base/directives/field.client.directive.js:4]: don't forget to remove

**Comment by liangyuanruo** [comment]
> Address edge case of query parameters before hashbang.
> 
> This is a known issue with angularjs $location service.
> Current thinking is to extract $location.absUrl() and parse using query-string

let's just keep it simple and stick with $location. we'll just tell the user that they must ascertain that the query parameters are correct.

> Address edge case of duplicated query parameter keys
> 
> Current thinking is to display both query param values in the field value, with a note: Multiple inputs with same key detected in URL.

don't mind if the algorithm simply overwrites using the last key-value pair

> Address case of special characters in query param values
> 
> Need to encode URI correctly?

yes, URI encoding must be retained. in fact - best to add tests for this as it's a key functionality of the redirect endpoint.

**Comment by liangyuanruo** [comment]
could you also check whether XSS is possible? we need to make sure `<script>...</script>` and HTML tags cannot be injected from the query parameters on the frontend.

you may wish to check out the [$sanitize](https://docs.angularjs.org/api/ngSanitize/service/$sanitize) and [$sce](https://docs.angularjs.org/api/ng/service/$sce) services.

File: src/app/controllers/public-forms.server.controller.js (hunk 2 above)
Comment (mantariksh) [reply src/app/controllers/public-forms.server.controller.js:55]: don't we need to support special characters as well? I think it's quite unexpected for the user if they type a comma or full stop and suddenly their prefill stops working

File: src/app/controllers/public-forms.server.controller.js (hunk 1 above)
Comment (shuli-ogp) [reply src/app/controllers/public-forms.server.controller.js:47]: @liangyuanruo considered this, but req.query returns an object, and that means order of the query params may not be retained

the implication is that upon redirect (i.e. adding the hashbang), the user who wants to eyeball the url to make sure it is still correct may find it difficult to compare with the expected string

if this is not a major concern, then will use req.query

Comment (shuli-ogp) [reply src/app/controllers/public-forms.server.controller.js:47]: actually on further consideration it's also less complicated to pass the query string as is to the redirect, with res.query we are reconstructing the url string from the query object - thoughts?

File: src/app/controllers/public-forms.server.controller.js (hunk 2 above)
Comment (shuli-ogp) [reply src/app/controllers/public-forms.server.controller.js]: got it, I've relaxed the validation for this

File: src/app/routes/frontend.server.routes.js (hunk 3 above)
Comment (shuli-ogp) [reply src/app/routes/frontend.server.routes.js]: as above, i've relaxed the validation for this

File: src/public/modules/forms/base/directives/field.client.directive.js (hunk 4 above)
Comment (shuli-ogp) [reply src/public/modules/forms/base/directives/field.client.directive.js]: removed

File: src/app/controllers/public-forms.server.controller.js (hunk 1 above)
Comment (liangyuanruo) [reply src/app/controllers/public-forms.server.controller.js:47]: query parameters isn't supposed to have inherent order as well, so this shouldn't be an issue

File: src/public/modules/forms/base/directives/field.client.directive.js [hunk 5]
```diff
@@ -23,6 +28,27 @@ function fieldDirective(FormFields) {
...
+
+      const queryParams = $location.search()
+      console.log(queryParams)
+
+      if (
+        scope.field._id in queryParams &&
+        scope.field.fieldType === 'textfield'
+      ) {
+        let prefillValue = queryParams[scope.field._id]
+        prefillValue = Array.isArray(prefillValue)
```
Comment (liangyuanruo) [inline src/public/modules/forms/base/directives/field.client.directive.js:44]: it's more defensive if we test whether the prefill value is a string, since that is the only use case we would like to support

File: tests/unit/backend/controllers/public-forms.server.controller.spec.js [hunk 6]
```diff
@@ -90,6 +91,21 @@ describe('Public-Forms Controller', () => {
...
+      req.url = '/321564654f65we4f65e4f5?abc=def&zzz=yyy'
+      const uriString = encodeURIComponent('abc=def&zzz=yyy')
+      res.redirect = jasmine.createSpy().and.callFake(() => {
+        expect(res.redirect).toHaveBeenCalledWith(
+          `/#!/321564654f65we4f65e4f5?${uriString}`,
+        )
+        done()
+      })
+      Controller.redirect(req, res)
+    })
```
Comment (karrui) [inline tests/unit/backend/controllers/public-forms.server.controller.spec.js:107]: Add more tests for more unicode characters rather than just ASCII. (Tamil, Bengali, Chinese, etc) so we can check if redirect behaviour is retained for those string types.

File: src/public/modules/forms/base/directives/field.client.directive.js [hunk 7]
```diff
@@ -23,6 +28,27 @@ function fieldDirective(FormFields) {
       isValidateDate: '<',
     },
     link: function (scope) {
+      // Stealth prefill feature
+      // If a query parameter is provided to a form URL in the form ?<fieldId1>=<value1>&<fieldId2>=<value2>...
+      // And if the fieldIds are valid mongoose object IDs and refer to a short text field,
+      // Then prefill and disable editing the corresponding form field on the frontend
+
+      const queryParams = $location.search()
+      console.log(queryParams)
```
Comment (karrui) [inline src/public/modules/forms/base/directives/field.client.directive.js:37]: Remove console log

File: src/public/modules/forms/base/directives/field.client.directive.js [hunk 8]
```diff
@@ -23,6 +28,27 @@ function fieldDirective(FormFields) {
...
+        scope.field._id in queryParams &&
+        scope.field.fieldType === 'textfield'
+      ) {
+        let prefillValue = queryParams[scope.field._id]
+        prefillValue = Array.isArray(prefillValue)
+          ? prefillValue[prefillValue.length - 1] // use last value in array if param key repeated
+          : // but note that there is no specification that the last value will definitely be last in order in the url string
+            prefillValue
+        scope.field.fieldValue = $sanitize(prefillValue) // sanitize prefillValue as a precaution, but this means "<>& cannot be used
+        scope.field.disabled = true
```
Comment (karrui) [inline src/public/modules/forms/base/directives/field.client.directive.js:49]: I thought the case was to prefill but leave it modifiable? What was the conclusion? @liangyuanruo @syan-syan

File: src/app/controllers/public-forms.server.controller.js (hunk 1 above)
Comment (karrui) [reply src/app/controllers/public-forms.server.controller.js:47]: Agree, just use `req.query`.

File: src/app/controllers/public-forms.server.controller.js [hunk 9]
```diff
@@ -44,9 +44,13 @@ exports.isFormPublic = function (req, res, next) {
...
+  const queryPortion = req.url ? req.url.split('?', 2)[1] : undefined
   let redirectPath = req.params.state
     ? req.params.Id + '/' + req.params.state
     : req.params.Id
+  if (queryPortion && queryPortion.length > 0) {
+    redirectPath = redirectPath + '?' + encodeURIComponent(queryPortion)
+  }
   try {
     const metaTags = await fetchMetatags(req)
     return res.render('index', {
```
Comment (karrui) [inline src/app/controllers/public-forms.server.controller.js:56]: Not related to this PR, but how does this code work? How does redirect work when we already return a render before the redirect?

Comment (karrui) [reply src/app/controllers/public-forms.server.controller.js:56]: Self answered, the final redirect is for when metatags fail to be created... should probably move it into the catch block with a return statement for clarity

**Comment by shuli-ogp** [comment]
@liangyuanruo 
tried and XSS doesn't seem possible, as the prefilled values are rendered as text, but will sanitise the input as a precaution
`fieldId1=<script>abc<%2Fscript>&fieldId2=<html>abc<%2Fhtml>`
Screenshot 2020-10-28 at 3 31 36 PM

File: src/public/modules/forms/base/directives/field.client.directive.js (hunk 8 above)
Comment (shuli-ogp) [reply src/public/modules/forms/base/directives/field.client.directive.js]: @syan-syan @karrui based on what liwei told me (intent is for public to scan the form to prefill a serial number) I think it should be non-modifiable, since the purpose is to avoid manual entry errors

File: src/app/controllers/public-forms.server.controller.js (hunk 9 above)
Comment (liangyuanruo) [reply src/app/controllers/public-forms.server.controller.js:57]: @karrui can file an issue for this? this PR needs to go out quickly

Comment (karrui) [reply src/app/controllers/public-forms.server.controller.js:57]: This doesn't affect functionality, all g. I'll fix it in my refactor

Comment (karrui) [reply src/app/controllers/public-forms.server.controller.js:57]: But I want to check, comments point to metatags only being created if it's a crawler, but I don't see any code checking if it's a crawler? Seems like all redirects create the metatags and render index regardless? Is it old code or?

File: tests/unit/backend/controllers/public-forms.server.controller.spec.js (hunk 6 above)
Comment (liangyuanruo) [reply tests/unit/backend/controllers/public-forms.server.controller.spec.js:107]: we can add the tests, but i just checked with Li Wei, and the first cut only requires alphanumeric, hyphen, and underscore characters.

File: src/public/modules/forms/base/directives/field.client.directive.js (hunk 8 above)
Comment (liangyuanruo) [reply src/public/modules/forms/base/directives/field.client.directive.js]: that is correct

File: src/app/controllers/public-forms.server.controller.js (hunk 9 above)
Comment (liangyuanruo) [reply src/app/controllers/public-forms.server.controller.js:57]: yeah, this is pretty old code that Kevan hastily wrote when he was briefly on forms.

File: src/public/modules/forms/base/directives/field.client.directive.js (hunk 7 above)
Comment (shuli-ogp) [reply src/public/modules/forms/base/directives/field.client.directive.js]: done

File: src/app/controllers/public-forms.server.controller.js (hunk 9 above)
Comment (shuli-ogp) [reply src/app/controllers/public-forms.server.controller.js:57]: done

File: src/public/modules/forms/base/directives/field.client.directive.js [hunk 10]
```diff
@@ -23,6 +25,28 @@ function fieldDirective(FormFields) {
...
+
+      if (
+        scope.field._id in queryParams &&
+        scope.field.fieldType === 'textfield'
+      ) {
+        const prefillValue = queryParams[scope.field._id]
+        if (typeof prefillValue === 'string') {
+          // Only support unique query params. If query params are duplicated,
+          // none of the duplicated keys will be prefilled
+          scope.field.fieldValue = prefillValue // do not use $sanitize as it removes non-english characters, tested that script and html tags do not work
```
Comment (liangyuanruo) [inline src/public/modules/forms/base/directives/field.client.directive.js:45]: to put in $sanitize

Comment (shuli-ogp) [reply src/public/modules/forms/base/directives/field.client.directive.js]: Reinserted

File: src/public/modules/forms/base/directives/field.client.directive.js [hunk 11]
```diff
@@ -23,6 +25,28 @@ function fieldDirective(FormFields) {
       isValidateDate: '<',
     },
     link: function (scope) {
+      // Stealth prefill feature
+      // If a query parameter is provided to a form URL in the form ?<fieldId1>=<value1>&<fieldId2>=<value2>...
+      // And if the fieldIds are valid mongoose object IDs and refer to a short text field,
+      // Then prefill and disable editing the corresponding form field on the frontend
+
+      const decodedUrl = he.decode($location.url()) // tech debt; after redirect, & is encoded as &amp; in the query string
+      const query = decodedUrl.split('?')[1]
```
Comment (liangyuanruo) [inline src/public/modules/forms/base/directives/field.client.directive.js:39]: can i confirm there's no null pointer exception if the URL doesn't contain `?`?

Comment (liangyuanruo) [reply src/public/modules/forms/base/directives/field.client.directive.js:39]: or index out of bounds with the `[1]` index access

**Review by liangyuanruo — APPROVED** [review]
LGTM, with a couple of questions remaining, including why hidden fields aren't being prefilled as there's no explicit check right now.

**Comment by liangyuanruo** [comment]
Note that it's possible to leave out a value in order to lock a field, although I'm hesitant to call this a bug outright as this could prove useful. The field would have to be optional for this to work, though.

image

File: src/public/modules/forms/base/directives/field.client.directive.js (hunk 11 above)
Comment (shuli-ogp) [reply src/public/modules/forms/base/directives/field.client.directive.js:39]: yes, there is no null pointer exception because when we call [split](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/split), `If separator is omitted or does not occur in str, the returned array contains one element consisting of the entire string.`

accessing index out of bounds returns undefined, but since this isn't in the specifications i've added an array length check

File: tests/unit/backend/controllers/public-forms.server.controller.spec.js (hunk 6 above)
Comment (shuli-ogp) [reply tests/unit/backend/controllers/public-forms.server.controller.spec.js:121]: i've added in the hyphen and underscore to tests just to be sure; will add the other unicode chars in the next cut when we support non-english characters

**Comment by shuli-ogp** [comment]
@liangyuanruo 

> why hidden fields aren't being prefilled as there's no explicit check right now.

i tested this again on staging, and wasn't able to replicate this and it works now (i.e. prefills for hidden fields do work). I may simply have missed out to hard-refresh when we were testing it this afternoon (i loaded the form without the query params first to retrieve the fieldIds), which resulted in the query params not being loaded

File: src/app/controllers/public-forms.server.controller.js [hunk 12]
```diff
@@ -62,8 +67,9 @@ exports.redirect = async function (req, res) {
       },
       error: err,
     })
+    res.redirect('/#!/' + redirectPath)
+    return
```
Comment (mantariksh) [inline src/app/controllers/public-forms.server.controller.js:71]: nit
```suggestion
    return res.redirect('/#!/' + redirectPath)
```

**Review by mantariksh — APPROVED** [review]
suggest the following test:
- [ ] Prefilled field value is cleared when field is hidden by logic at the point of submission
